### ISSUE TITLE
Memory usage is high with OpenSSL1.1.1a

### ISSUE THREAD

**Issue by vgmahajanshetty** [openssl/openssl#8869]
Hi,

I am working on load balancer. I am observing that with OpenSSL1.1.1a the memory usage is really high compared to OpenSSL1.0.2.

We are running as server in this particular case; with OpenSSL1.1.1a memory usage of the server goes up to 10GB when we run with 3,00,000 concurrent clients (connections).
When we run same server with OpenSSL1.0.2 with same number of clients, memory usage is about 4GB. So there is increase of about 2.5 times with OpenSSL1.1.1a.

We are negotiating TLS1.2 in both cases.

I just checked size of some SSL structures, looks like they have grown up in size with OpenSSL1.1.1a.
In OpenSSL1.1.1a:
(gdb) p sizeof(SSL)
$1 = 6232
(gdb) p sizeof(SSL_CTX)
$2 = 1024
(gdb) p sizeof(SSL_SESSION)
$3 = 696

In OpenSSL1.0.2 sizes are as below:
(gdb) p sizeof(SSL)
$1 = 824
(gdb)  p sizeof(SSL_CTX)
$2 = 800
(gdb) p sizeof(SSL_SESSION)
$3 = 352

We can see there is big difference in size for SSL struct.

Is there any way to reduce it (through compilation flags or some other means) at least when we don't want to turn on TLS1.3?

Thank you.

**Comment by mattcaswell** [comment]
Looking at the sizeof the structs isn't a very reliable measure of how size has changed over time since there are many things which are stored as pointers to some other object inside these structs - and some things may have moved from being pointers to being directly in the structs, or vice versa (so the memory usage remains the same but the sizeof value changes). Additionally I generally wouldn't expect any sizeof differences in SSL_CTX to have a significant impact since typically you have many SSL objects for one SSL_CTX (or even all your SSL object for just one SSL_CTX).

That said it is certainly true that there are some significant extra fields added in order to support TLSv1.3. These are allocated regardless of whether you end up negotiating TLSv1.3. Unfortunately these fields are not compiled out by "no-tls1_3" so that wouldn't have an impact. You might try compiling by passing "-DOPENSSL_SMALL_FOOTPRINT" to Configure to see if that makes any difference.

**Comment by vgmahajanshetty** [comment]
Hi Matt,

Thanks for your response.
I understand that size of process has not increased only just because of increase in structs size, it might be one of the reason; or may not be the reason at all as some pointers might have become direct members as you mentioned. Just for quick reference I pasted them.

As you mentioned, I will try with -DOPENSSL_SMALL_FOOTPRINT option and check if it can make any difference.

Thanks,
Vishwanath M

**Comment by vgmahajanshetty** [comment]
Hi,

I checked with the OpenSSL compiled by passing -DOPENSSL_SMALL_FOOTPRINT, I am not finding any difference. Memory utilization is same as it was earlier.
Looks like there is no way to reduce memory usage; please let me know if I can try any other options.

Thanks,
Vishwanath G M

**Comment by mattcaswell** [comment]
I have no further suggestions. We might want to look in the future to see if there is anything we can do to reduce the TLSv1.3 memory burden.

**Comment by vgmahajanshetty** [comment]
Ok, just to clarify again: issue is not only with TLSv1.3 negotiation, it's with TLSv1.2 also.

Thank You.

**Comment by mattcaswell** [comment]
> Ok, just to clarify again: issue is not only with TLSv1.3 negotiation, it's with TLSv1.2 also.

Yes, but the problem comes with the TLSv1.3 code which consumes memory in the SSL object regardless of whether TLSv1.3 is negotiated.

**Comment by richsalz** [comment]
This issue is very vague.  What is "too high"?  Compared to previous releases? How much memory per connection? Are resources properly free'd? Have you used a memory-checker to see which objects are consuming the most space?

**Comment by vgmahajanshetty** [comment]
Hi,

I havn't used mem-checker. Can you please let me know or point me a link which tells about how to use mem-checker? I will try that.

Thanks.

**Comment by FdaSilvaYY** [comment]
See my previous answer to this question [here](https://github.com/openssl/openssl/issues/8078#issuecomment-457003930).

**Comment by nhorman** [comment]
marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
